src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -230,37 +259,12 @@ async ValueTask DecrementInFlightCountAsync(InFlightSolution inFlightSolution)
         {
             try
             {
-                // Try to create the solution snapshot incrementally off of the workspaces CurrentSolution first.
-                var updater = new SolutionCreator(Services.HostServices, assetProvider, this.CurrentSolution);
-                if (await updater.IsIncrementalUpdateAsync(newSolutionChecksum, cancellationToken).ConfigureAwait(false))
-                {
-                    return await updater.CreateSolutionAsync(newSolutionChecksum, cancellationToken).ConfigureAwait(false);
-                }
-                else
-                {
-                    // Otherwise, this is a different solution, or the first time we're creating this solution.  Bulk
-                    // sync over all assets for it.
-                    await assetProvider.SynchronizeSolutionAssetsAsync(newSolutionChecksum, cancellationToken).ConfigureAwait(false);
-
-                    // get new solution info and options
-                    var solutionInfo = await assetProvider.CreateSolutionInfoAsync(newSolutionChecksum, cancellationToken).ConfigureAwait(false);
-                    var solution = CreateSolutionFromInfo(solutionInfo);
+                var solutionToUpdate = await GetOrCreateSolutionToUpdateAsync(
+                    assetProvider, newSolutionChecksum, cancellationToken).ConfigureAwait(false);
 
-                    // ensure that the solution has the correct source generator execution versions. note we should do
-                    // this all in a unified fashion with CreateSolutionAsync above.  However, that is blocked in
-                    // https://github.com/dotnet/roslyn/pull/72860
-                    {
-                        var newSolutionCompilationChecksums = await assetProvider.GetAssetAsync<SolutionCompilationStateChecksums>(
-                            AssetPathKind.SolutionCompilationStateChecksums, newSolutionChecksum, cancellationToken).ConfigureAwait(false);
-
-                        var newVersions = await assetProvider.GetAssetAsync<SourceGeneratorExecutionVersionMap>(
-                            AssetPathKind.SolutionSourceGeneratorExecutionVersionMap, newSolutionCompilationChecksums.SourceGeneratorExecutionVersionMap, cancellationToken).ConfigureAwait(false);
-
-                        solution = solution.UpdateSpecificSourceGeneratorExecutionVersions(newVersions);
-                    }
```

Comment:
note: this logic intentionally disappears.  because we now just have it fall out from calling into the normal CreateSolutionAsync copdepath which does the appropraite checksum diffing/syncing.  (It also does a validation sanity check in debug, which this codepath never did).

Reply:
Nice!